controller/snapshot_controller.go

```diff
@@ -0,0 +1,585 @@
+		UpdateFunc: sc.enqueueEngineChange,
+	}, 0)
+
+	return sc
+}
+
+func (sc *SnapshotController) enqueueSnapshot(obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", obj, err))
+		return
+	}
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return
+	}
+	if namespace != sc.namespace {
+		return
+	}
+
+	snap, err := sc.ds.GetSnapshot(name)
+	if err != nil {
+		sc.queue.Add(key)
+		return
+	}
+
+	vol, err := sc.ds.GetVolumeRO(snap.Spec.Volume)
+	if err != nil {
+		sc.queue.Add(key)
+		return
+	}
+
+	if vol.Status.OwnerID != sc.controllerID {
+		return
+	}
+
+	sc.queue.Add(key)
+}
+
+func (sc *SnapshotController) enqueueEngineChange(oldObj, curObj interface{}) {
+	curEngine, ok := curObj.(*longhorn.Engine)
+	if !ok {
+		deletedState, ok := curObj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("received unexpected obj: %#v", curObj))
+			return
+		}
+		// use the last known state, to enqueue, dependent objects
+		curEngine, ok = deletedState.Obj.(*longhorn.Engine)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("DeletedFinalStateUnknown contained invalid object: %#v", deletedState.Obj))
+			return
+		}
+	}
+
+	vol, err := sc.ds.GetVolumeRO(curEngine.Spec.VolumeName)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("snapshot controller failed to get volume %v when enqueuing engine %v: %v", curEngine.Spec.VolumeName, curEngine.Name, err))
+	}
+
+	if vol.Status.OwnerID != sc.controllerID {
+		return
+	}
+
+	oldEngine, ok := oldObj.(*longhorn.Engine)
+	if !ok {
+		deletedState, ok := oldObj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("received unexpected obj: %#v", oldObj))
+			return
+		}
+		// use the last known state, to enqueue, dependent objects
+		oldEngine, ok = deletedState.Obj.(*longhorn.Engine)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("DeletedFinalStateUnknown contained invalid object: %#v", deletedState.Obj))
+			return
+		}
+	}
+
+	needEnqueueSnapshots := curEngine.Status.CurrentState != oldEngine.Status.CurrentState ||
+		!reflect.DeepEqual(curEngine.Status.PurgeStatus, oldEngine.Status.PurgeStatus) ||
+		!reflect.DeepEqual(curEngine.Status.Snapshots, oldEngine.Status.Snapshots)
+
+	if !needEnqueueSnapshots {
+		return
+	}
+
+	snapshots, err := sc.ds.ListVolumeSnapshotsRO(curEngine.Spec.VolumeName)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to list snapshots for volume %v when enqueuing engine %v: %v", curEngine.Spec.VolumeName, oldEngine.Name, err))
+		return
+	}
+
+	snapshots = getSnapshotsForEnqueuingFromEngines(oldEngine, curEngine, snapshots)
+	for _, snap := range snapshots {
+		sc.enqueueSnapshot(snap)
+	}
+
+	return
+}
+
+func getSnapshotsForEnqueuingFromEngines(oldEngine, curEngine *longhorn.Engine, snapshots map[string]*longhorn.Snapshot) map[string]*longhorn.Snapshot {
```

Comment:
NIT: rename this. e.g. `filterSnapshotsForEngineEnqueuing`

Reply:
Thank you. Fixed